File: frontend/src/types/candidate.ts
/**
 * TypeScript interfaces matching the backend candidate schema.
 * Mirrors candidate_schema.json and backend/app/schemas/candidate.py.
 */

export type ProficiencyLevel = "beginner" | "intermediate" | "advanced" | "expert";

export interface Skill {
  skill_name: string;
  proficiency: ProficiencyLevel;
  duration_months: number;
  endorsements: number;
  skill_assessment_score?: number | null;
}

export interface CareerHistory {
  company_name: string;
  title: string;
  start_date: string; // ISO date string e.g. "2020-01"
  end_date: string | null; // null = current
  duration_months: number;
  description: string;
  is_current: boolean;
}

export interface Education {
  institution: string;
  degree: string;
  field_of_study: string;
  start_year: number;
  end_year: number | null;
}

export interface Certification {
  name: string;
  issuer: string;
  issue_date: string;
  expiry_date: string | null;
}

export interface Language {
  language: string;
  proficiency: string;
}

export interface RedrobSignals {
  open_to_work_flag: boolean;
  last_active_date: string; // ISO datetime
  notice_period_days: number;
  recruiter_response_rate: number; // 0.0–1.0
  avg_response_time_hours: number;
  verified_email: boolean;
  verified_phone: boolean;
  linkedin_connected: boolean;
  github_activity_score: number; // -1 if not linked
  saved_by_recruiters_30d: number;
  profile_completeness_score: number; // 0–100
  willing_to_relocate: boolean;
}

export interface Candidate {
  candidate_id: string; // format: CAND_XXXXXXX
  headline: string;
  summary: string;
  current_title: string;
  current_company: string;
  years_of_experience: number;
  location: string;
  expected_salary_range_inr_lpa: {
    min: number;
    max: number;
  };
  skills: Skill[];
  career_history: CareerHistory[];
  education: Education[];
  certifications: Certification[];
  languages: Language[];
  redrob_signals: RedrobSignals;
  created_at?: string;
  updated_at?: string;
}

export interface CandidateListResponse {
  items: Candidate[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

File: backend/app/schemas/candidate.py
"""Pydantic v2 schemas for Candidate and related models."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ── CareerHistory ─────────────────────────────────────────────────────────────

class CareerHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    duration_months: Optional[int] = None
    is_current: bool = False
    company_size: Optional[str] = None
    location: Optional[str] = None


# ── Education ─────────────────────────────────────────────────────────────────

class EducationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    institution: Optional[str] = None
    degree: Optional[str] = None
    field_of_study: Optional[str] = None
    start_year: Optional[int] = None
    end_year: Optional[int] = None
    grade: Optional[str] = None


# ── Skill ─────────────────────────────────────────────────────────────────────

class SkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    proficiency: Optional[str] = None
    duration_months: Optional[int] = None
    endorsements: int = 0


# ── Certification ─────────────────────────────────────────────────────────────

class CertificationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    issuing_organization: Optional[str] = None
    issue_date: Optional[date] = None
    expiry_date: Optional[date] = None
    credential_id: Optional[str] = None


# ── Language ──────────────────────────────────────────────────────────────────

class LanguageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    proficiency: Optional[str] = None


# ── Candidate ─────────────────────────────────────────────────────────────────

class CandidateBase(BaseModel):
    """Shared fields used by both Create and Read schemas."""

    candidate_id: str = Field(..., description="Source dataset identifier, e.g. CAND_0001234")
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    headline: Optional[str] = None
    summary: Optional[str] = None
    current_title: Optional[str] = None
    current_company: Optional[str] = None
    location: Optional[str] = None
    country: Optional[str] = None
    years_of_experience: Optional[float] = None
    preferred_work_mode: Optional[str] = None
    willing_to_relocate: Optional[bool] = None


class CandidateCreate(CandidateBase):
    """Schema for creating a new Candidate record."""
    pass


class CandidateRead(CandidateBase):
    """Full candidate read response including all nested relationships."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    career_history: List[CareerHistoryRead] = []
    education: List[EducationRead] = []
    skills: List[SkillRead] = []
    certifications: List[CertificationRead] = []
    languages: List[LanguageRead] = []


class CandidateListItem(BaseModel):
    """Lightweight candidate item for list views (no nested data)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    candidate_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    current_title: Optional[str] = None
    current_company: Optional[str] = None
    location: Optional[str] = None
    years_of_experience: Optional[float] = None
